org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -16,63 +16,106 @@ namespace MixedReality.Toolkit.Input
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                previouslyDetectedCollider.GetComponent<IXRProximityInteractable>().OnProximityExited(new ProximityExitedEventArgs(this, previouslyDetectedCollider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Remove(previouslyDetectedCollider);
-                    }
+                    currentlyDetectedInteracable.OnProximityEntered(new ProximityHoverEnteredEventArgs(currentlyDetectedInteracable));
                 }
-                foreach (Collider collider in DetectedColliders)
+            }
+        }
+
+        /// <summary>
+        /// Returns a hashset of all unique interactables that are currently detected.
+        /// </summary>
+        /// <returns>Hashset with unique interactables currently detected</returns>
+        private void UpdateCurrentlyDetectedInteractables()
+        {
+            currentlyDetectedInteractables.Clear();
+
+            foreach (Collider collider in DetectedColliders)
+            {
+                if (InteractionManager.TryGetInteractableForCollider(collider, out IXRInteractable xrInteractable) &&
+                    xrInteractable is IXRProximityInteractable xrProximityInteractable &&
+                    !currentlyDetectedInteractables.Contains(xrProximityInteractable))
                 {
-                    if (!previouslyDetectedColliders.Contains(collider))
-                    {
-                        IXRProximityInteractable nearInteractionMode = collider.GetComponent<IXRProximityInteractable>();
-                        if (nearInteractionMode != null)
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                collider.GetComponent<IXRProximityInteractable>().OnProximityEntered(new ProximityEnteredEventArgs(this, collider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Add(collider);
-                    }
+                    currentlyDetectedInteractables.Add(xrProximityInteractable);
                 }
             }
-            return result;
         }
 
+        /// <summary>
+        /// Indicates whether any near interactable has interactors selecting.
+        /// </summary>
+        /// <returns>True if an interactable has interactors selecting, false otherwise.</returns>
         private bool IsNearInteractorSelecting()
         {
-            foreach (XRBaseInteractor nearInteractor in nearInteractors)
+            foreach (XRBaseInteractable nearInteractable in nearInteractables)
             {
-                if (nearInteractor.hasSelection)
+                if (nearInteractable.interactorsSelecting.Count > 0)
```

Comment:
There's no guarantee that the interactor selecting this interactable has to be a near interactor. Many interactors can be selected both by near and far interaction, even simultaneously.

Reply:
Reverted in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/ee2f619a385491f73d7303e544ff944704e0042f, resolving.